docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,255 @@
+
+The compactor’s _DeletedSeriesCleaner_ will apply this logic on individual blocks and each time it is run, it creates a new block without the data that matched the deletion request. The original individual blocks containing the data that was requested to be deleted, need to be marked for deletion by the compactor.
+
+While deleting the data permanently from the block storage, the `meta.json` files will be used to keep track of the deletion progress. Inside each `meta.json` file, we will add a new field called `tombstonesFiltered`. This will store an array of deletion request id's that were used to create this block. Once the rewrite logic is applied to a block, the new block's `meta.json` file will append the deletion request id(s) used for the rewrite operation inside this field. This will let the _DeletedSeriesCleaner_ know that this block has already processed the particular deletions requests listed in this field. Assuming that the deletion requests are quite rare, the size of the meta.json files should remain small.
+
+The _DeletedSeriesCleaner_ can iterate through all the blocks that the deletion request could apply to. For each of these blocks, if the deletion request ID isn't inside the meta.json `tombstonesFiltered` field, then the compactor can apply the rewrite logic to this block. If there are multiple tombstones that are currently being processing for deletions and apply to a particular block, then the _DeletedSeriesCleaner_ will process both at the same time to prevent additional blocks from being created. If after iterating through all the blocks, it doesn’t find any such blocks requiring deletion, then the `Pending` state is complete and the request progresses to the `Processed` state.
+
+One important thing to note regarding this rewrite tool is that it should not be used at the same time as when another compactor is touching a block. If the tool is run at the same time as compaction on a particular block, it can cause overlap and the data marked for deletion can already be part of the compacted block. To mitigate such issues, these are some of the proposed solutions:
+
+Option 1: Only apply the deletion once the blocks are in the final state of compaction.
+
+Pros:
+- Simpler implementation as everything is contained within the DeletedSeriesCleaner.
+
+Cons:
+- Might have to wait for a longer period of time for the compaction to be finished.
+    - This would mean the earliest time to be able to run the deletion would be once the last time from the block_ranges in the [compactor_config](https://cortexmetrics.io/docs/blocks-storage/compactor/#compactor-configuration) has passed. By default this value is 24 hours, so only once 24 hours have passed and the new compacted blocks have been created, then the rewrite can be safely run.
```

Comment:
This means that if we configure a larger compaction block duration for example 7d, then we must wait for 7 days for the tombstones to be cleaned up.
From Thanos project perspective this will be a little bit long as Thanos's max block is 14d time range.

Reply:
Yeah, if the block duration is longer, then with this deletion design, it will require a longer wait. 

Maybe for the Thanos project the design could be expanded to perform deletions during compaction. If you have other alternatives in mind, I would be happy to discuss with you.